src/vm/virtualMachine.ts

```diff
@@ -125,381 +56,292 @@ export class VirtualMachine {
-    const op0Addr =
-      op0Register === Register.Fp
-        ? this.fp.add(op0Offset)
-        : this.ap.add(op0Offset);
-
-    let op0 = this.memory.get(op0Addr);
-
-    // Compute the second operand address based on the op1Source and
-    // the offset
-    const op1Addr = this.computeOp1Address(op1Source, op1Offset, op0);
-    let op1 = this.memory.get(op1Addr);
-
-    // If op0 is undefined, then we can deduce it from the instruction, dst and op1
-    // We also deduce the result based on the result logic
-    if (op0 === undefined) {
-      const deducedOp0 = this.deduceOp0(instruction, dst, op1);
+    const registers = {
+      [Register.Pc]: this.pc,
+      [Register.Ap]: this.ap,
+      [Register.Fp]: this.fp,
+    };
 
-      if (deducedOp0 !== undefined) {
-        this.memory.assertEq(op0Addr, deducedOp0);
-      }
-      op0 = deducedOp0;
-    }
+    const op0Addr: Relocatable = registers[op0Register].add(op0Offset);
+    const op1Addr: Relocatable = registers[op1Register].add(op1Offset);
+    const dstAddr: Relocatable = registers[dstRegister].add(dstOffset);
 
-    // If operand 1 is undefined, then we can deduce it from the instruction,
-    // destination and operand 0.
-    // We also deduce the result based on the result logic
-    if (op1 === undefined) {
-      const deducedOp1 = this.deduceOp1(instruction, dst, op0);
+    let op0: SegmentValue | undefined = this.memory.get(op0Addr);
+    let op1: SegmentValue | undefined = this.memory.get(op1Addr);
+    let res: SegmentValue | undefined = undefined;
+    let dst: SegmentValue | undefined = this.memory.get(dstAddr);
 
-      if (deducedOp1 !== undefined) {
-        this.memory.assertEq(op1Addr, deducedOp1);
-      }
-    }
+    switch (opcode | resLogic) {
+      case Opcode.Call | ResLogic.Op1:
+        ({ op0, op1 } = this.validateOperandsCallOpcode(instruction, op0, op1));
+        res = op1;
+        dst = this.fp;
+        break;
 
-    // If res is undefined, then we can compute it from op0 and op1
-    if (res === undefined) {
-      const computedRes = this.computeRes(
-        instruction,
-        op0 as SegmentValue,
-        op1 as SegmentValue
-      );
+      case Opcode.Call | ResLogic.Add:
+        ({ op0, op1 } = this.validateOperandsCallOpcode(instruction, op0, op1));
+        res = op0.add(op1);
+        dst = this.fp;
+        break;
 
-      res = computedRes;
-    }
+      case Opcode.Call | ResLogic.Mul:
+        ({ op0, op1 } = this.validateOperandsCallOpcode(instruction, op0, op1));
+        if (!isFelt(op0)) throw new ExpectedFelt();
+        res = op0.mul(op1);
+        dst = this.fp;
+        break;
 
-    // If dst is undefined, then we can deduce it from the instruction and res
-    if (dst === undefined) {
-      const deducedDst = this.deduceDst(instruction, res);
+      case Opcode.Call | ResLogic.Unused:
+        ({ op0, op1 } = this.validateOperandsCallOpcode(instruction, op0, op1));
+        res = undefined;
```

Comment:
`res` is already `undefined` no?

Reply:
Indeed, same goes for the other case with ResLogic.Unused (in the default case). I wanted to be explicit but as the switch is directly after the declaration of res (to undefined), it is useless --> to be removed